### PR TITLE
Add external gcs connection support

### PR DESCRIPTION
**Pull request by johnsonshih**

### CODE REVIEW

File: vm.go [hunk 1]
```diff
@@ -139,9 +153,21 @@ func CreateVirtualMachineSpec(opts *VirtualMachineOptions) (*VirtualMachineSpec,
...
+				UseVsock:            opts.GuestConnectionUseVsock,
+				UseConnectedSuspend: true,
+			}
+		} else {
+			// gcs connects to hcsshim (external)
+			// Allow administrators and SYSTEM to bind to vsock sockets
+			// so that we can create GCS sockets.
+			spec.VirtualMachine.Devices.HvSocket = &hcsschema.HvSocket2{
+				HvSocketConfig: &hcsschema.HvSocketSystemConfig{
+					DefaultBindSecurityDescriptor: "D:P(A;;FA;;;SY)(A;;FA;;;BA)",
```
Comment (jagadishmurugan) [inline vm.go:168]: where do we get this value? is this documented? #Resolved

File: vm.go [hunk 2]
```diff
@@ -701,6 +769,82 @@ func (vm *VirtualMachineSpec) String() string {
...
+		return conn, err
+	case <-ctx.Done():
+	}
+	l.Close()
+	err := <-ch
+	if err == nil {
+		return conn, err
+	}
+
+	// Prefer context error to VM error to accept error in order to return the
```
Comment (jagadishmurugan) [inline vm.go:850]: nit: rephrase comment - something like below:
Prefer following return order of error: Context error over VM error...

File: vm.go [hunk 3]
```diff
@@ -701,6 +769,82 @@ func (vm *VirtualMachineSpec) String() string {
...
+
+func (vm *VirtualMachineSpec) listenVsock(port uint32) (net.Listener, error) {
+	return winio.ListenHvsock(&winio.HvsockAddr{
+		VMID:      vm.runtimeId,
+		ServiceID: winio.VsockServiceID(port),
+	})
+}
+
+// acceptAndClose accepts a connection and then closes a listener. If the
+// context becomes done or the VM terminates, the operation will be
```
Comment (jagadishmurugan) [inline vm.go:827]: nit: context becomes done -> does it mean "context completed or available"? #Resolved

File: vm.go [hunk 4]
```diff
@@ -701,6 +769,82 @@ func (vm *VirtualMachineSpec) String() string {
...
+					log.G(ctx).WithError(err).Error("failed to roll back resource add")
+				}
+			}
+		}()
+	}
+	err = vm.gc.Modify(ctx, doc.GuestRequest)
+	if err != nil {
+		return fmt.Errorf("guest modify: %s", err)
+	}
+	if doc.ResourcePath != "" && doc.RequestType == requesttype.Remove {
```
Comment (jagadishmurugan) [inline vm.go:808]: should we use a else or switch case? operationtype can be either Add or Remove... 
unless "err = vm.gc.Modify(ctx, doc.GuestRequest)" has to be done prior to handling Remove... #Resolved

File: vm.go (hunk 1 above)
Comment (johnsonshih) [reply vm.go:168]: It's acl in string format, meaning of content is expressed in comment "to Allow administrators and SYSTEM to bind to vsock sockets"

File: vm.go (hunk 4 above)
Comment (johnsonshih) [reply vm.go:808]: we only need to do this part for remove.  For remove, guest requests are served first and then host requests.

File: vm.go (hunk 3 above)
Comment (johnsonshih) [reply vm.go:827]: context becomes done means "ctx timeout or cancelled"

caller provides context, so context is always available, we don't do anything to modify context, so context won't completed in this function or go function

File: vm.go (hunk 4 above)
Comment (jagadishmurugan) [reply vm.go:808]: can u confirm the following:
     Add => Host Request followed by Guest Request 
     Remove => Guest Request followed by Host Request

Comment (johnsonshih) [reply vm.go:808]: if no guest request or external gcs is not used, invoke hcs to handle the request.

if with guest request and external gcs is used:
For Add, process host request and then guest request
For Remove, process guest request and then host request
For others, just handle quest request

**Review by jagadishmurugan — APPROVED** [review]

**Pull request merged**